flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/CheckpointBarrierUnaligner.java

```diff
@@ -154,46 +154,51 @@ public void processBarrier(CheckpointBarrier receivedBarrier, int channelIndex)
 	@Override
 	public void processCancellationBarrier(CancelCheckpointMarker cancelBarrier) throws Exception {
 		long cancelledId = cancelBarrier.getCheckpointId();
-		if (LOG.isDebugEnabled()) {
-			LOG.debug("{}: Checkpoint {} canceled, aborting alignment.", taskName, cancelledId);
-		}
-
-		if (currentConsumedCheckpointId >= cancelledId && !isCheckpointPending()) {
-			return;
-		}
+		// tag whether we should abort checkpoint from task thread view
+		boolean shouldAbort1 = false;
 
-		if (isCheckpointPending()) {
+		if (cancelledId > currentConsumedCheckpointId) {
+			currentConsumedCheckpointId = cancelledId;
+			shouldAbort1 = true;
+		} else if (cancelledId == currentConsumedCheckpointId && isCheckpointPending()) {
 			LOG.warn("{}: Received cancellation barrier for checkpoint {} before completing current checkpoint {}. " +
 					"Skipping current checkpoint.",
 				taskName,
 				cancelledId,
 				currentConsumedCheckpointId);
 
 			resetConsumedBarriers();
+			shouldAbort1 = true;
+		}
+
+		// tag whether we should abort checkpoint from threadSafeUnaligner view
+		boolean shouldAbort2 = threadSafeUnaligner.setCancelledCheckpointId(cancelledId);
```

Comment:
I don't fully understand this `boolean shouldAbort2`. What are the conditions when it's set to true while `shouldAbort1` is false? Why do we have those two sources of truth?

Reply:
The checkpoint might happen either by netty thread via `notifyBarrierReceived` or by task thread via `processBarrier`, and we are not sure which one would happen earlier in the race condition, so we need to compare both `currentReceivedCheckpointId` and `currentConsumedCheckpointId` with canceled id to abort checkpoint properly.

If `shouldAbort2` true and `shouldAbort1` false, that means the `notifyBarrierReceived` triggered by netty thread happen earlier. E.g. the netty thread receives the ch2 from channel1, and the task thread processes the ch1 from channel2, then the task thread processes the cancellation ch2 from channel3. In this case, `shouldAbort1` will be false and `shouldAbort2` will be true.

No matter which tag is true, we should abort the checkpoint.

Reply:
> If shouldAbort2 true and shouldAbort1 false, that means the notifyBarrierReceived triggered by netty thread happen earlier.

But this `notifyBarrierReceived` from netty thread, would enqueue mailbox action. 

If the mailbox action hasn't yet been executed, we could mark this checkpoint as aborted here, in this method (task thread), and prevent mail from executing (task thread).

If the mailbox action has already executed (task thread), it could have left the `currentConsumedCheckpointId` field already up to date, so this method would already know whether to abort or not from `shouldAbort1`.

Or is it about aborting the checkpoint after `notifyBarrierReceived` from netty thread, but before enqueued mailbox action was executed?

Reply:
> If the mailbox action hasn't yet been executed, we could mark this checkpoint as aborted here, in this method (task thread), and prevent mail from executing (task thread).

Yes, we already did it in this PR. If the cancelled checkpoint id is larger than current id, the current id would be updated both in `CheckpointBarrierUnaligner` and `ThreadSafeUnaligner`. And when the triggered checkpoint prepares to execute from mailbox action, it would be exit directly by comparing the triggered id with current id.

> If the mailbox action has already executed (task thread), it could have left the currentConsumedCheckpointId field already up to date, so this method would already know whether to abort or not from shouldAbort1.

Not alway the case. If the checkpoint 1 triggered by netty thread and then executed by task thread, that does not mean the task thread already processed the respective triggered barrier in advance, so the `currentConsumedCheckpointId` might still out of date when checkpoint 1 was executed. 

In the async process of checkpointing, the task thread might continue processing from other channels, then it might process cancellation checkpoint to abort the current executing checkpoint.

> Or is it about aborting the checkpoint after notifyBarrierReceived from netty thread, but before enqueued mailbox action was executed?

Yes, it can cover this case as well. All in all, it can abort the un-executed checkpoint which was enqueued into mailbox already and also abort the going checkpoint which might be in async process.

Reply:
The key problem here is that we actually maintain two suits of states in `CheckpointBarrierUnaligner` and `ThreadSafeUnaligner` separately, and we are not sure which states are up to date in race condition case. Therefore we need to judge both of them to give a final decision.